Support for API Keys

Temporal is introducing support for api keys , in addition to mTLS , https://docs.temporal.io/cloud/api-keys

Adding this support will make it easier to get started .

Hi...this didn't work. I get the following exception:

```
istening for transport dt_socket at address: 5005
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2024-11-20 16:59:13,175 INFO  [io.tem.ser.WorkflowServiceStubsImpl] (Quarkus Main Thread) Created WorkflowServiceStubs for channel: ManagedChannelOrphanWrapper{delegate=ManagedChannelImpl{logId=1, target=ap-south-1.aws.api.temporal.io:7233}}

2024-11-20 16:59:19,425 ERROR [io.qua.run.Application] (Quarkus Main Thread) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:121)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:116)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:268)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:249)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:167)
	at io.temporal.api.workflowservice.v1.WorkflowServiceGrpc$WorkflowServiceBlockingStub.getSystemInfo(WorkflowServiceGrpc.java:4864)
	at io.temporal.serviceclient.SystemInfoInterceptor.getServerCapabilitiesOrThrow(SystemInfoInterceptor.java:132)
	at io.temporal.serviceclient.SystemInfoInterceptor.lambda$getServerCapabilitiesWithRetryOrThrow$0(SystemInfoInterceptor.java:118)
	at io.temporal.internal.retryer.GrpcSyncRetryer.retry(GrpcSyncRetryer.java:69)
	at io.temporal.internal.retryer.GrpcRetryer.retryWithResult(GrpcRetryer.java:60)
	at io.temporal.serviceclient.SystemInfoInterceptor.getServerCapabilitiesWithRetryOrThrow(SystemInfoInterceptor.java:117)
	at io.temporal.serviceclient.ChannelManager.lambda$getServerCapabilities$3(ChannelManager.java:346)
	at io.temporal.internal.retryer.GrpcRetryer.retryWithResult(GrpcRetryer.java:60)
	at io.temporal.serviceclient.ChannelManager.connect(ChannelManager.java:320)
	at io.temporal.serviceclient.WorkflowServiceStubsImpl.connect(WorkflowServiceStubsImpl.java:161)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.temporal.internal.WorkflowThreadMarker.lambda$protectFromWorkflowThread$1(WorkflowThreadMarker.java:83)
	at jdk.proxy3/jdk.proxy3.$Proxy60.connect(Unknown Source)
	at io.temporal.serviceclient.WorkflowServiceStubs_9w2blsdkS3yzcHs-FT2nR_13XeA_Synthetic_ClientProxy.connect(Unknown Source)
	at io.temporal.worker.WorkerFactory.start(WorkerFactory.java:213)
	at io.quarkiverse.temporal.WorkerFactoryRecorder.startWorkerFactory(WorkerFactoryRecorder.java:123)
	at io.quarkus.deployment.steps.TemporalProcessor$startWorkers1590318771.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.TemporalProcessor$startWorkers1590318771.deploy(Unknown Source)
	... 11 more
Caused by: java.net.SocketException: Connection reset
	at java.base/sun.nio.ch.SocketChannelImpl.throwConnectionReset(SocketChannelImpl.java:401)
	at java.base/sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:434)
	at io.grpc.netty.shaded.io.netty.buffer.PooledByteBuf.setBytes(PooledByteBuf.java:254)
	at io.grpc.netty.shaded.io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1132)
	at io.grpc.netty.shaded.io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:357)
	at io.grpc.netty.shaded.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:151)
	at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.grpc.netty.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.grpc.netty.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more


2024-11-20 16:59:19,450 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure
2024-11-20 16:59:19,450 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure

```


The configuration is as follows:
```
quarkus.temporal.namespace=<redacted>
quarkus.temporal.connection.target=ap-south-1.aws.api.temporal.io:7233
quarkus.temporal.connection.api-key=<redacted>
```

The setup at the temporal side is correct, since if i use the java code as below, the connection is established.

```
  @GET
  @Path("/spike")
  public void start() throws SSLException {
    // Create a Metadata object with the Temporal namespace header key.
    Metadata.Key<String> TEMPORAL_NAMESPACE_HEADER_KEY = Metadata.Key.of("temporal-namespace",
        Metadata.ASCII_STRING_MARSHALLER);
    Metadata metadata = new Metadata();
    metadata.put(TEMPORAL_NAMESPACE_HEADER_KEY, "<redacted>");

    // Create the Workflow service stub.
    WorkflowServiceStubsOptions.Builder stubOptions = WorkflowServiceStubsOptions.newBuilder()
        .setChannelInitializer(
            (channel) -> {
              channel.intercept(MetadataUtils.newAttachHeadersInterceptor(metadata));
            })
        .addGrpcMetadataProvider(
            new AuthorizationGrpcMetadataProvider(() -> "Bearer "
                + "<redacted>"))
        .setTarget("ap-south-1.aws.api.temporal.io:7233");
    stubOptions.setSslContext(SimpleSslContextBuilder.noKeyOrCertChain().setUseInsecureTrustManager(false).build());

    WorkflowServiceStubs service = WorkflowServiceStubs.newServiceStubs(stubOptions.build());

    /*
     * Get a Workflow service client which can be used to start, Signal, and Query
     * Workflow Executions.
     */
    WorkflowClient client = WorkflowClient.newInstance(
        service, WorkflowClientOptions.newBuilder().setNamespace("<redacted>").build());
    
  }
```

--- a/extension/runtime/src/main/java/io/quarkiverse/temporal/WorkflowServiceStubsRecorder.java
+++ b/extension/runtime/src/main/java/io/quarkiverse/temporal/WorkflowServiceStubsRecorder.java
@@ -82,7 +82,14 @@ public class WorkflowServiceStubsRecorder {
                     (channel) -> {
                         channel.intercept(MetadataUtils.newAttachHeadersInterceptor(metadata));
                     });
-            builder.addGrpcMetadataProvider(new AuthorizationGrpcMetadataProvider(() -> "Bearer " + connection.apiKey()));
+            builder.addGrpcMetadataProvider(
+                    new AuthorizationGrpcMetadataProvider(() -> "Bearer " + connection.apiKey().get()));
+            try {
+                builder.setSslContext(
+                        SimpleSslContextBuilder.noKeyOrCertChain().setUseInsecureTrustManager(false).build());
+            } catch (SSLException e) {
+                throw new ConfigurationException("Failed to create SSL context", e);
+            }
         } else {
             // Mutual Transport Layer Security
             MTLSRuntimeConfig mtls = connection.mtls();
@@ -117,7 +124,8 @@ public class WorkflowServiceStubsRecorder {
 
         WorkflowServiceStubsOptions.Builder builder = WorkflowServiceStubsOptions.newBuilder()
                 .setChannel(
-                        (ManagedChannel) context.getInjectedReference(Channel.class, GrpcClient.Literal.of("temporal-client")))
+                        (ManagedChannel) context.getInjectedReference(Channel.class,
+                                GrpcClient.Literal.of("temporal-client")))
                 .setMetricsScope(createScope(isMicrometerEnabled))
                 .setRpcRetryOptions(createRpcRetryOptions(connection.rpcRetry()));
 
@@ -165,7 +173,8 @@ public class WorkflowServiceStubsRecorder {
     /**
      * Read the content of the path.
      * <p>
-     * The file is read from the classpath if it exists, otherwise it is read from the file system.
+     * The file is read from the classpath if it exists, otherwise it is read from
+     * the file system.
      *
      * @param path the path, must not be {@code null}
      * @return the content of the file

